base/number.jl

```diff
@@ -136,6 +136,45 @@ true
 """
 signbit(x::Real) = x < 0
 
+"""
+    ispositive(x)
+
+Test whether `x` is positive. Falls back to [`isnan`](@ref), [`signbit`](@ref) and [`iszero`](@ref). See also [`isnegative`](@ref).
+
+# Examples
+```jldoctest
+julia> ispositive(-4.0)
+false
+
+julia> ispositive(99)
+true
+
+julia> ispositive(0.0)
+false
+```
+"""
+ispositive(x) = !isnan(x) && !signbit(x) && !iszero(x)
+
+"""
+    isnegative(x)
+
+Test whether `x` is positive. Falls back to [`isnan`](@ref), [`signbit`](@ref) and [`iszero`](@ref). See also [`ispositive`](@ref).
+
+# Examples
+```jldoctest
+julia> isnegative(-4.0)
+true
+
+julia> isnegative(99)
+false
+
+julia> isnegative(-0.0)
+false
+```
+"""
+isnegative(x) = !isnan(x) && signbit(x) && !iszero(x)
+isnegative(x::Integer) = signbit(x) # skip iszero
```

Comment:
Nothing about `Integer` guarantees that `-0` isn't a valid number. That property is specific to `Base.BitInteger` and `BigInt`. I don't think this specialization is valid.

Reply:
You could also argue that integer can be `Inf` and `NaN` as well (e.g. Infinities.jl), but they are not handled by the Julia base. Do you have an example of integer `-0`?

Reply:
No. But I could go create `SignedMagnitudeIntegers.jl` today and have a type with `-0`. See [Signed magnitude representation](https://en.wikipedia.org/wiki/Signed_number_representations#Sign%E2%80%93magnitude).

Reply:
The fallback for `signbit(::Integer)` is https://github.com/JuliaLang/julia/blob/a9611ce6acde605c4f5ec5787514e181b6b4add6/base/int.jl#L139

so I'd argue that my implementation is consistent. The drawback is that if you have a type like that, then you need to specialize `isnegative` and `signbit` at the same time, so basically one more task (compared to only `signbit`).
The advantage is that for integer types that don't have `-0`, one can either have a shortcut for `signbit` which makes `isnegative` faster, or it will work identically as `isnegative(x) = x<0`.